terragrunt/modules/docs-rs/README.md

```diff
@@ -0,0 +1,62 @@
+# `docs-rs`
+
+This is the [Terraform] module that defines the infrastructure for [docs.rs].
+
+**NOTE**: As of Feb 2023, this infrastructure configuration is only used for the docs.rs staging environment. The production environment running on docs.rs is using a different architecture. The plan of record is to move production to using this configuration.
+
+## Overview
+
+[docs.rs] consists of a few different components:
+
+  - A web frontend hosted as a container running in an [ECS] cluster.
+  - A background documentation builder running on [ec2](https://aws.amazon.com/ec2) instances as part of an [autoscaling](https://aws.amazon.com/ec2/autoscaling) group.
+  - A crate registry watcher (**TODO**: this is currently not configured).
+
+These are documented in more detail below.
+
+Data is stored in two different locations:
+
+  - Static rendered crate documentation is stored in [S3](https://aws.amazon.com/s3).
+  - Other data is stored in a postgresql database [RDS](https://aws.amazon.com/rds) instance.
+
+### The Flow
+
+```mermaid
+flowchart TD
+    web[Web Frontend]<-->Postgres
+    s3-->web
+    worker[Background Worker]<-->Postgres
+    worker-->s3
+    watcher[Registry Watcher]-->Postgres
+```
+
+The registry watcher watches the registry for new entries. When a new crate is added, it adds that crate to a queue stored in the database. The background worker watches the queue and works off crates from the queue as quickly as possible. It builds crate documentation and stores the artifacts in s3, adds relevant information to the database, and removes the crate from the queue. The web frontend then displays that information to the user.
+
+## Web Frontend
+
+```mermaid
+flowchart LR
+    DNS1[Frontend DNS] --> CloudFront
+    CloudFront --> DNS2[ECS LB DNS]
+    DNS2 --> ECSLB[Cluster Load Balancer]
+    ECSLB --> cluster[ECS Cluster]
+    cluster --> service[ECS Service]
+    service --> task[ECS Task]
+    task --> Container
+```
```

Comment:
I've been experimenting a little bit with the flowchart. Would be nice to have the request cycle as a straight line at the center and the other nodes on the top or at the bottom, but I haven't figured out if that's even possible with Mermaid.

```mermaid
flowchart LR
    crates.io -. DNS ..-> CloudFront
    ecs.docs-rs-prod.rust-lang.net -. DNS .-> ECSLB

    User -- Request--> CloudFront
    CloudFront --> ECSLB[Application Load Balancer]    
    ECSLB --> service[ECS Service]
    
    subgraph ecs [ECS Cluster]
        service --> Container
        service -.-> task[Task Definition]
    end
```

Reply:
Wow! Your Mermaid foo is *much* better than mine...